src/core/dbus-manager.c

```diff
@@ -2783,44 +2783,83 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
-        Manager *m = ASSERT_PTR(userdata);
+        _cleanup_free_ char *prefix = NULL;
+        Unit *caller;
+        pid_t pid;
         uid_t user;
         gid_t group;
         int r;
 
         assert(message);
+        assert(manager);
+        assert(ret_user);
+        assert(ret_group);
 
-        if (!MANAGER_IS_SYSTEM(m))
-                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED,
-                                         "This method is only supported on the system manager.");
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID|SD_BUS_CREDS_UID|SD_BUS_CREDS_GID|SD_BUS_CREDS_AUGMENT, &creds);
+        if (r < 0)
+                return r;
 
-        r = mac_selinux_access_check(message, "acquire-namespace", error);
+        r = sd_bus_creds_get_pid(creds, &pid);
         if (r < 0)
                 return r;
 
-        r = bus_verify_acquire_namespace_async(m, message, error);
+        r = sd_bus_creds_get_uid(creds, &user);
         if (r < 0)
                 return r;
-        if (r == 0)
-                return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
 
-        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_UID|SD_BUS_CREDS_GID|SD_BUS_CREDS_AUGMENT, &creds);
+        r = sd_bus_creds_get_uid(creds, &group);
         if (r < 0)
                 return r;
 
-        r = sd_bus_creds_get_uid(creds, &user);
+        *ret_user = user;
+        *ret_group = group;
+
+        /* After this point, an error simply means the caller is not a user manager instance, so return false. */
+
+        caller = manager_get_unit_by_pid(manager, pid);
+        if (!caller || !unit_name_is_valid(caller->id, UNIT_NAME_INSTANCE))
+                return 0;
+
+        if (unit_name_to_prefix(caller->id, &prefix) < 0)
+                return 0;
+
+        return streq(prefix, "user");
+}
+
+/* Create a new user namespace that maps all UIDs/GIDs in the system and return the
+ * file descriptor to the caller. */
+static int method_get_identity_mapping(sd_bus_message *message, void *userdata, sd_bus_error *error) {
+        _cleanup_free_ char *uid_map = NULL, *gid_map = NULL;
+        _cleanup_close_ int fd = -1;
+        Manager *m = ASSERT_PTR(userdata);
+        uid_t user;
+        gid_t group;
+        int r;
+
+        assert(message);
+
+        if (!MANAGER_IS_SYSTEM(m))
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED,
+                                         "This method is only supported on the system manager.");
+
+        r = mac_selinux_access_check(message, "acquire-namespace", error);
         if (r < 0)
                 return r;
 
-        r = sd_bus_creds_get_gid(creds, &group);
+        /* User manager instances are allowed to bypass Polkit, and are always authorized as they will use this to set up
+         * sandboxing options for user units. Everything else goes through the Polkit auth step first. */
```

Comment:
sorry, but we can't do that for security decision like that. pids can be recycled, and thus we can read stuff from /proc/$PID for decorative purposes, but not to make security decisions on. what's the rationale for trying to avoid PK lookups here?

Reply:
is there any other way to go from a dbus message to the systemd unit that it originates from?